ehrql/serializer.py

```diff
@@ -0,0 +1,265 @@
+from ehrql.query_model.table_schema import BaseConstraint, Column, TableSchema
+from ehrql.utils.module_utils import get_all_subclasses
+
+
+# Mapping of names to types for all types that we support serializing/deserializing
+TYPE_REGISTRY = {
+    type_.__qualname__: type_
+    for type_ in [
+        bool,
+        int,
+        float,
+        str,
+        datetime.date,
+        tuple,
+        dict,
+        frozenset,
+        Position,
+        TableSchema,
+        Column,
+        pathlib.Path,
+        ColumnSpec,
+        DummyDataConfig,
+        Measure,
+        *get_all_subclasses(Node),
+        *get_all_subclasses(BaseCode),
+        *get_all_subclasses(BaseConstraint),
+        *get_all_subclasses(BaseDatasetReader),
+    ]
+}
+
+TYPE_REGISTRY_INVERSE = {type_: name for name, type_ in TYPE_REGISTRY.items()}
+
+# Ensure no name clashes
+assert len(TYPE_REGISTRY) == len(TYPE_REGISTRY_INVERSE)
+
+
+def type_name(obj):
+    return TYPE_REGISTRY_INVERSE[type(obj)]
+
+
+def serialize(value):
+    return json.dumps(Marshaller.to_dict(value), indent=2)
+
+
+def deserialize(data):
+    return Unmarshaller.from_dict(json.loads(data))
+
+
+class Marshaller:
+    """
+    Convert arbitrary values to nested structures of JSON-supported values ready to be
+    serialized
+    """
+
+    @classmethod
+    def to_dict(cls, value):
+        marshaller = cls()
+        return {
+            "value": marshaller.marshal(value),
+            **marshaller.get_references(),
+        }
+
+    def __init__(self):
+        self.references = {}
+        # Sequence of unique IDs (strings not ints so they can be used as JSON keys)
+        id_sequence = (str(i) for i in range(1, 2**32))  # pragma: no branch
+        self.get_next_reference_id = id_sequence.__next__
+
+    def get_references(self):
+        return dict(self.references.values())
+
+    @functools.singledispatchmethod
+    def marshal(self, obj):
+        assert False, f"Unsupported type '{type(obj)}': {obj!r}"
+
+    @marshal.register(type(None))
+    @marshal.register(bool)
+    @marshal.register(str)
+    @marshal.register(int)
+    @marshal.register(float)
+    def marshal_primitive(self, obj):
+        return obj
+
+    @marshal.register(datetime.date)
+    def marshal_date(self, obj):
+        return {type_name(obj): obj.isoformat()}
+
+    @marshal.register(list)
+    def marshal_list(self, obj):
+        return [self.marshal(value) for value in obj]
+
+    @marshal.register(tuple)
+    @marshal.register(frozenset)
+    def marshal_iterable(self, obj):
+        return {type_name(obj): [self.marshal(value) for value in obj]}
+
+    @marshal.register(dict)
+    def marshal_dict(self, obj):
+        return {
+            type_name(obj): [
+                [self.marshal(key), self.marshal(value)] for key, value in obj.items()
+            ],
+        }
+
+    @marshal.register(type)
+    def marshal_type(self, obj):
+        return {"type": TYPE_REGISTRY_INVERSE[obj]}
```

Comment:
It doesn't seem obvious to me that the set of types of objects that we want to be able to serialize is the same as the set of _types_ that we want to be able to serialize. Is there a justification, or is it just a pragmatic implementation based on the latter being fairly obviously a subset of the former?

Reply:
They're obviously not the same but, as you say, the latter is necessarily a subset of the former as the types we need to serialize are just those which a Series might need to contain, and therefore we also need to be able to serialize instances of those types.

And I can't see the harm in the serializer being able to serialize references to types instead of just instances of those types, even for the types where we don't strictly need to support that.

Reply:
No harm to the implementation. Tiny harm to comprehensibility -- when I read this I wonder if I'm missing some reason for this _necessarily_ being the same list. Maybe worth a tiny comment?